core/payment-driver/src/gnt/common.rs

```diff
@@ -0,0 +1,165 @@
+    ))
+}
+
+pub(crate) async fn get_gnt_balance(
+    gnt_contract: &Contract<Http>,
+    address: Address,
+) -> PaymentDriverResult<Balance> {
+    gnt_contract
+        .query("balanceOf", (address,), None, Options::default(), None)
+        .compat()
+        .await
+        .map_or_else(
+            |e| Err(PaymentDriverError::LibraryError(format!("{:?}", e))),
+            |amount| {
+                Ok(Balance::new(
+                    utils::u256_to_big_dec(amount)?,
+                    Currency::Gnt {},
+                ))
+            },
+        )
+}
+
+pub(crate) fn verify_gnt_tx<T: Transport>(
+    receipt: &TransactionReceipt,
+    contract: &Contract<T>,
+) -> PaymentDriverResult<()> {
+    verify_gnt_tx_logs(&receipt.logs, contract)?;
+    verify_gnt_tx_status(&receipt.status)?;
+    Ok(())
+}
+
+fn verify_gnt_tx_status(status: &Option<U64>) -> PaymentDriverResult<()> {
+    match status {
+        None => Err(PaymentDriverError::UnknownTransaction),
+        Some(status) => {
+            if *status == U64::from(config::ETH_TX_SUCCESS) {
+                Ok(())
+            } else {
+                Err(PaymentDriverError::FailedTransaction)
+            }
+        }
+    }
+}
+
+fn verify_gnt_tx_logs<T: Transport>(
+    logs: &Vec<Log>,
+    contract: &Contract<T>,
+) -> PaymentDriverResult<()> {
+    if logs.len() != config::TRANSFER_LOGS_LENGTH {
+        return Err(PaymentDriverError::UnknownTransaction);
+    }
+    verify_gnt_tx_log(&logs[0], contract)?;
+    Ok(())
+}
+
+fn verify_gnt_tx_log<T: Transport>(log: &Log, contract: &Contract<T>) -> PaymentDriverResult<()> {
+    if log.address != contract.address() {
+        return Err(PaymentDriverError::UnknownTransaction);
+    }
+    verify_gnt_tx_log_topics(&log.topics)?;
+    verify_gnt_tx_log_data(&log.data)?;
+    Ok(())
+}
+
+fn verify_gnt_tx_log_topics(topics: &Vec<H256>) -> PaymentDriverResult<()> {
+    if topics.len() != config::TX_LOG_TOPICS_LENGTH {
+        return Err(PaymentDriverError::UnknownTransaction);
+    }
+    // topics[0] is the keccak-256 of the Transfer(address,address,uint256) canonical signature
+    verify_gnt_tx_log_canonical_signature(&topics[0])?;
+    Ok(())
+}
+
+fn verify_gnt_tx_log_canonical_signature(canonical_signature: &H256) -> PaymentDriverResult<()> {
+    if *canonical_signature
+        != H256::from_slice(&hex::decode(config::TRANSFER_CANONICAL_SIGNATURE).unwrap())
+    {
+        return Err(PaymentDriverError::UnknownTransaction);
+    }
+    Ok(())
+}
+
+fn verify_gnt_tx_log_data(data: &Bytes) -> PaymentDriverResult<()> {
+    if data.0.len() != config::TX_LOG_DATA_LENGTH {
+        return Err(PaymentDriverError::UnknownTransaction);
+    }
+    Ok(())
+}
+
+pub(crate) fn build_payment_details(
+    receipt: &TransactionReceipt,
+) -> PaymentDriverResult<PaymentDetails> {
+    // topics[1] is the value of the _from address as H256
+    let sender = utils::topic_to_address(&receipt.logs[0].topics[1]);
```

Comment:
should we check the `&receipt.logs[0].topics` length before doing all these hard-coded indexes?

Reply:
Before building payment details it is verified by 
```rust
pub(crate) fn verify_gnt_tx<T: Transport>
```